### PR TITLE
(feat): New OCR3 Consensus Capability Aggregator: Reduce

### PR DESCRIPTION
**Pull request by justinkaseman**
### Description
Take multiple observations and reduce down to a single report (Values.Map).
Intended to be flexible enough to cover 95% of single report use cases by allowing workflow authors to define how they would like their data points aggregated, what format it outputs as, along with different deviation strategies.

More complex aggregation, such as spawning multiple reports, would require a different feature for accepting a user supplied aggregator.

### CODE REVIEW

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/values/pb"
+)
+
+type AggregatorConfig struct {
+	// Static data that is added into every report
+	AdditionalData map[string]any `mapstructure:"additionalData"`
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:25]: Is this going to be eg. the FeedID ?

Comment (cedric-cordenier) [reply pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:25]: I think if other fields are provided we should still try and aggregate them.

I see the config being slighltly different:
- a user provides a list of fields to aggregate along with its aggregation method.
  eg. FeedID - Mode, Price - Median, Timestamp - Median
- they also provide a field that we want to check the deviation logic against; in our case Price

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+			if oldValue == nil || currDeviation > field.Deviation.InexactFloat64() {
+				shouldReport = true
+			}
+		}
+	}
+
+	// Add static additonal data to the report
+	for key, value := range a.config.AdditionalData {
+		report[key] = value
+	}
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:101]: See my comment about additional data; we should still be aggregating this IMO

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+type AggregatorConfig struct {
+	// Static data that is added into every report
+	AdditionalData map[string]any `mapstructure:"additionalData"`
+	// Configuration on how to aggregate one or more data points
+	Fields []AggregationField `mapstructure:"fields"`
+	// The top level field name that report data is put into
+	OutputFieldName string `mapstructure:"outputFieldName" json:"outputFieldName" default:"Reports"` // TODO: is this actually needed?
+	// The structure surrounding the report data that is put on to "OutputFieldName"
+	ReportFormat string `mapstructure:"reportFormat" json:"reportFormat" default:"map" jsonschema:"enum=map,enum=array"`
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:31]: Why is this needed? Feels like extra complexity, but I may well be missing the point :/

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+	}
+	a.lggr.Debugw("current state initialized", "state", currentState, "previousOutcome", previousOutcome)
+	return &currentState, nil
+}
+
+func (a *reduceAggregator) extractValues(observations map[ocrcommon.OracleID][]values.Value, aggregationKey string) []any {
+	events := map[string][]any{}
+	for nodeID, nodeObservations := range observations {
+		// TODO: check the following comment:
+		// we only expect a single observation per node
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:172]: I think this is fine for the MVP 👍

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+		vmap, err := nodeObservations[0].Unwrap()
+		if err != nil {
+			a.lggr.Warnf("could not unwrap observation value from node %d: %v", nodeID, err)
+			continue
+		}
+
+		value := vmap.(map[string]any)[aggregationKey]
+
+		switch v := value.(type) {
+		case string:
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:190]: @justinkaseman I wonder if you'd be better off dealing with the type complexity as low down as possible, rather than unwrapping the types here and dealing with it again further down.

Concretely, you just pass the values.Value to either mode or median and let those functions figure out what types they support + how they aggregate.

Eg. for mode, you can just do what the identical aggregator does and marshal the proto to some bytes and sha it. For median, you can just convert each type to a decimal (I think that's what we agreed on anyway in our last chat) and return an error if any one of the values can't be handled.

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go (hunk 3 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:31]: It _is_ extra complexity that I don't love having in -- but the purpose is:

The Data Feeds consumer expects that EncodableOutcome looks like:
```
{
"Reports": [
      {
         "FeedID": []byte,
         "Price": *big.Int,
         "Timestamp": int64
      },
...
}
```

Notice that it's an array of reports, so that it can handle multiple mercury reports coming through.

The EVM Encoder gets the config:
image

So this Aggregator config is to tell it to wrap in an array, even if there's only one report.

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go (hunk 5 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:190]: That does seem cleaner. I'll try that.

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go (hunk 1 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:25]: Yes exactly, you can see the CCIP Data Feeds use case recreated in [the uncommented test](https://github.com/smartcontractkit/chainlink-common/pull/842/files#diff-e3c476aac2b96c73893bd0d49244f6bc32b63e97c98c5876e6368000712209b8R210).

Yeahhh, that's fair - just run it all through aggregation. I had thought about that too. I'll make that change.

File: pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go (hunk 3 above)
Comment (cedric-cordenier) [reply pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go:31]: Ah I see, sorry I missed that initially. Makes sense 👍

File: pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json [hunk 6]
```diff
@@ -35,7 +35,8 @@
         "aggregation_method": {
           "type": "string",
           "enum": [
-            "data_feeds"
+            "data_feeds",
+            "reduce"
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json:39]: @justinkaseman I don't think this is correct -- or at least it's not what we intended for the consensus capability: we expect each aggregation method to have its own config + struct, rather than extending the existing data feeds consensus one.

For identical consensus, I created this struct here: https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/capabilities/consensus/ocr3/ocr3cap/identical_consensus.go, you'll need to create an equivalent for Reduce consensus.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+		if len(vals) < 2*f+1 {
+			return nil, fmt.Errorf("not enough observations provided %s, have %d want %d", field.InputKey, len(vals), 2*f+1)
+		}
+
+		singleValue, err := reduce(field.Method, vals)
+		if err != nil {
+			return nil, fmt.Errorf("unable to reduce on method %s, err: %s", field.Method, err.Error())
+		}
+
+		if field.DeviationType != "none" {
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:85]: Can we put this in a constant, along with the other deviation types :)

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 8]
```diff
@@ -0,0 +1,436 @@
...
+	DeviationType string `mapstructure:"deviationType" json:"deviationType,omitempty" jsonschema:"enum=percent,enum=absolute,enum=none"`
+}
+
+type reduceAggregator struct {
+	config ReduceAggConfig
+}
+
+var _ types.Aggregator = (*reduceAggregator)(nil)
+
+func (a *reduceAggregator) Aggregate(lggr logger.Logger, previousOutcome *types.AggregationOutcome, observations map[ocrcommon.OracleID][]values.Value, f int) (*types.AggregationOutcome, error) {
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:59]: This can be a follow up, but can we add some debug logging throughout to record what's happening during the aggregation step? Eg. current/target deviation and heartbeat values, along with the final shouldReport decision and the executionID.

From experience not having this makes it really hard to debug what's going on, and it just means we end up adding it later.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+	// The key that the aggregated data is put under
+	// If omitted, the InputKey will be used
+	OutputKey string `mapstructure:"outputKey" json:"outputKey"`
+	// How the data set should be aggregated
+	Method string `mapstructure:"method" json:"method" jsonschema:"enum=median,enum=mode" required:"true"`
+	// An optional check to only report when the difference from the previous report exceeds a certain threshold.
+	// Can only be used when the field is of a numeric type: string, decimal, int64, big.Int, time.Time, float64
+	// If no deviation is provided on any field, there will always be a report once minimum observations are reached.
+	DeviationString string          `mapstructure:"deviation"  json:"deviation,omitempty"`
+	Deviation       decimal.Decimal `mapstructure:"-"  json:",omitempty"`
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:46]: Are we missing heartbeat here?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+		if err != nil {
+			return nil, err
+		}
+		err = mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	zeroValue, _ := values.Wrap(decimal.NewFromInt(0))
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:162]: Hmmmm... I don't think the result is the same here if the function errors vs if it doesn't, so I would still check the error, or better yet, manually construct the value:

```
var val values.Value
val = values.NewDecimal(decimal.NewFromInt(0))
```

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+			continue
+		}
+
+		val, err := nodeObservations[0].Unwrap()
+		if err != nil {
+			lggr.Warnf("node %d contributed a Value that could not be unwrapped", nodeID)
+			continue
+		}
+
+		// extract the value from complex types and re-wrap them
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:192]: A better comment here would explain why we need to do this: i.e. what feature of the value/aggregation is making it necessary to do this :)

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+			} else {
+				lggr.Warnf("aggregation key %s provided, but value is not an indexable type", aggregationKey)
+			}
+		}
+	}
+
+	return vals
+}
+
+func reduce(method string, items []values.Value) (values.Value, error) {
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:231]: Much cleaner :) Well done!

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+	// validations & fill defaults
+	if len(parsedConfig.Fields) == 0 {
+		return ReduceAggConfig{}, errors.New("reduce aggregator must contain config for Fields to aggregate")
+	}
+	for i, field := range parsedConfig.Fields {
+		if len(field.DeviationType) == 0 {
+			field.DeviationType = "none"
+			parsedConfig.Fields[i].DeviationType = "none"
+		}
+		if field.DeviationType != "absolute" && field.DeviationType != "percent" && field.DeviationType != "none" {
```
Comment (cedric-cordenier) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:405]: IMO this (and deviationType below) could use a helper:

eg.

```
isOneOf(field.DeviationType, []string{"percent", "absolute", "none"})
```

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 7 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:85]: Yes!

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 9 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:46]: This can be used as heartbeat by using an absolute deviation on the timestamp. I'll get some tests up to show that off.

File: pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json (hunk 6 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json:39]: Oops, yeah this shouldn't be there. Good catch. I'll make the new schema.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 11 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:192]: Expanding this comment makes me realize that I could just pass `any` type around. Keeping it as a Value doesn't really give any benefit. Where it's used for mode to get the protobuf I could just wrap it there.

File: pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json (hunk 6 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json]: Like so? https://github.com/smartcontractkit/chainlink-common/pull/842/commits/2076c0f23f5e9107043d705ad5a5512f4aab82c2

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 11 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go]: May not be a problem to move it around as a value though :grin:

Comment (cedric-cordenier) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go]: Yeah that makes sense :)

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 9 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:64]: Yes exactly right on the config for PoR.

Agreed on needing more documentation. I'll create a design doc for this.

File: pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json (hunk 6 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/ocr3cap/ocr3cap_data_feeds_consensus-schema.json]: Ahh, I see what you mean. Added.

**Review by cedric-cordenier — DISMISSED** [review]

File: pkg/capabilities/consensus/ocr3/models.go [hunk 14]
```diff
@@ -5,7 +5,7 @@ import (
 )
 
 type config struct {
-	AggregationMethod string      `mapstructure:"aggregation_method" json:"aggregation_method" jsonschema:"enum=data_feeds"`
+	AggregationMethod string      `mapstructure:"aggregation_method" json:"aggregation_method" jsonschema:"enum=data_feeds,enum=reduce"`
```
Comment (bolekk) [inline pkg/capabilities/consensus/ocr3/models.go:8]: Is this syntax correct? Not "enum=a,b,c"?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 15]
```diff
@@ -0,0 +1,504 @@
...
+	ocrcommon "github.com/smartcontractkit/libocr/commontypes"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/capabilities/consensus/ocr3/types"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink-common/pkg/values/pb"
+)
+
+const (
+	AGGREGATION_METHOD_MEDIAN = "median"
```
Comment (bolekk) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:25]: Maybe add a comment that it is not a true median (no averaging of two middle values)

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 16]
```diff
@@ -0,0 +1,504 @@
...
+		}
+		err = mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, fmt.Errorf("initializeCurrentState FromMapValueProto error: %s", err.Error())
+		}
+	}
+
+	zeroValue := values.NewDecimal(decimal.Zero)
+	for _, field := range a.config.Fields {
+		if _, ok := currentState[field.InputKey]; !ok {
```
Comment (bolekk) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:204]: what if InputKey is an empty string?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 17]
```diff
@@ -0,0 +1,504 @@
...
+			return nil, fmt.Errorf("not enough observations provided %s, have %d want %d", field.InputKey, len(vals), 2*f+1)
+		}
+
+		singleValue, err := reduce(field.Method, vals)
+		if err != nil {
+			return nil, fmt.Errorf("unable to reduce on method %s, err: %s", field.Method, err.Error())
+		}
+
+		if field.DeviationType != DEVIATION_TYPE_NONE {
+			oldValue := (*currentState)[field.InputKey]
```
Comment (bolekk) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:107]: What if InputKey is an empty string?

File: pkg/capabilities/consensus/ocr3/models.go (hunk 14 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/models.go:8]: It's the way that I see it in the example from json schema parser that we use
https://github.com/invopop/jsonschema?tab=readme-ov-file#example

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 16 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:204]: It gets set to
```
{
  "": value
}
```

But that's a good edge case in that if there are two Fields with no input key then they will clash.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 18]
```diff
@@ -0,0 +1,504 @@
...
+	SubMapField bool `mapstructure:"subMapField"  json:"subMapField,omitempty"`
+}
+
+type reduceAggregator struct {
+	config ReduceAggConfig
+}
+
+var _ types.Aggregator = (*reduceAggregator)(nil)
+
+func (a *reduceAggregator) Aggregate(lggr logger.Logger, previousOutcome *types.AggregationOutcome, observations map[ocrcommon.OracleID][]values.Value, f int) (*types.AggregationOutcome, error) {
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:80]: docstring on Aggregate

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 17 above)
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:107]: _nit_ personally this syntax is a bit rare/strange, might be nicer to just define a `Get(string) (any, error)` method on the current state's type

File: pkg/capabilities/consensus/ocr3/aggregators/identical.go [hunk 19]
```diff
@@ -14,10 +14,10 @@ import (
 )
 
 type identicalAggregator struct {
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/identical.go:16]: _nit_ touched, would be nice to add docstring of what is an `identicalAggregator`

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go [hunk 20]
```diff
@@ -0,0 +1,928 @@
...
+					mockValue, err := values.Wrap(int64(100))
+					require.NoError(t, err)
+					return map[commontypes.OracleID][]values.Value{1: {mockValue}, 2: {mockValue}, 3: {mockValue}}
+				},
+				shouldReport: true,
+				expectedOutcome: map[string]any{
+					"Reports": []any{
+						map[string]any{
+							"": int64(100),
+						},
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:59]: why is this a valid report, we don't want to error instead?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go [hunk 21]
```diff
@@ -0,0 +1,928 @@
...
+							"Price": true,
+						},
+					},
+				},
+				expectedState: map[string]any{
+					"BenchmarkPrice": true,
+				},
+			},
+			{
+				name: "aggregate on non-indexable type",
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:337]: what is being tested here exactly and why is this not a failure mode?  a report should reduce even if the observations are not a map?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go [hunk 22]
```diff
@@ -0,0 +1,928 @@
...
+							"Price": big.NewInt(100),
+						},
+					},
+				},
+				expectedState: map[string]any{
+					"BenchmarkPrice": big.NewInt(100),
+				},
+			},
+			{
+				name: "buffer mode",
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:238]: what is buffer mode and how would this work if the method was `median`? is that valid?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go [hunk 23]
```diff
@@ -0,0 +1,928 @@
...
+				},
+				shouldReport: true,
+				expectedOutcome: map[string]any{
+					"Reports": []any{
+						map[string]any{
+							"Price": int64(1),
+						},
+					},
+				},
+				expectedState: map[string]any{"": int64(1)},
```
Comment (MStreet3) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:359]: what is `expectedState` vs `expectedOutcome`, as in why doesn't this have `Price` as a key?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go (hunk 22 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:238]: Finding a mode on a data set of buffer / []byte types. Taking `median` of buffers isn't valid, because you can't sort them using a numeric comparison (`>`). We could support them by sorting them in another way though.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go (hunk 23 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:359]: `expectedState` is how the data is stored within the `metadata`, which is what gets passed in on every new aggregation round as what the "previous round" was.

`expectedOutcome` is what gets put as the workflow step's `result`, which currently is what ends up going to the chain-writer, and then on-chain.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 24]
```diff
@@ -0,0 +1,504 @@
...
+
+		(*currentState)[field.InputKey] = singleValue
+		if len(field.OutputKey) > 0 {
+			report[field.OutputKey] = singleValue
+		} else {
+			report[field.InputKey] = singleValue
+		}
+	}
+
+	// if SubMapKey is provided, move fields in a nested map
```
[the remark this answers is not in the document]
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:126]: It was a feature request for: data can nested data can be in the report, which enables the sub-piece to stay encoded as bytes in the contract, allowing storing it in the registry without knowing how it is decoded. The consumer from the registry will decode it all the way.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go (hunk 21 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:337]: Exactly, since the input to a capability is a `values.Value`, it could be either a List, Map, or just a non-complex type. We could either reject anything that isn't a Map, or map/list -- but to keep this flexible to future use cases, the code currently allows all of the above to come in.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 15 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:25]: Good call out. Added words to AggregationField struct.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go (hunk 20 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:59]: Yes, I think you are right about this being an invalid report. It shouldn't be a map with an empty key, rather just a non-complex value.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 16 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:204]: Since `{"": value}` shouldn't be a valid report, I added a new `reportFormat` for `value` which covers outputting a single value, and I changed it to now hold state as the output structure.

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go (hunk 22 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_test.go:238]: Renamed to bytes to make it more clear

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 17 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:107]: See above, same change should cover this

File: pkg/capabilities/consensus/ocr3/aggregators/identical.go (hunk 19 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/identical.go:16]: @bolekk how would you say `identicalAggregator` is different? One difference that I see is that it can enforce observations length. But maybe we can roll these together?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 25]
```diff
@@ -0,0 +1,523 @@
...
+	// If no deviation is provided on any field, there will always be a report once minimum observations are reached.
+	Deviation       decimal.Decimal `mapstructure:"-"  json:"-"`
+	DeviationString string          `mapstructure:"deviation"  json:"deviation,omitempty"`
+	// The format of the deviation being provided
+	// * percent - a percentage deviation
+	// * absolute - an unsigned numeric difference
+	DeviationType string `mapstructure:"deviationType" json:"deviationType,omitempty" jsonschema:"enum=percent,enum=absolute,enum=none"`
+	// The key to find a data point within the input data
+	// If omitted, the entire input will be used
+	InputKey string `mapstructure:"inputKey" json:"inputKey"`
```
Comment (ettec) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:62]: is this a key or a path? ie for nested structures you will get map of maps, maybe Path would be a better name?

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go [hunk 26]
```diff
@@ -0,0 +1,523 @@
...
+	case DEVIATION_TYPE_ABSOLUTE:
+		return diff, nil
+	case DEVIATION_TYPE_PERCENT:
+		if prevDeci.Cmp(decimal.NewFromInt(0)) == 0 {
+			if diff.Cmp(decimal.NewFromInt(0)) == 0 {
+				return decimal.NewFromInt(0), nil
+			}
+			return decimal.NewFromInt(math.MaxInt), nil
+		}
+		return diff.Div(prevDeci), nil
```
Comment (ettec) [inline pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:414]: should there be a zero check here or is it not possible for prevDeci to be 0?

Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go:414]: The zero check is right above there :grin: 

```
		if prevDeci.Cmp(decimal.NewFromInt(0)) == 0 {
			if diff.Cmp(decimal.NewFromInt(0)) == 0 {
				return decimal.NewFromInt(0), nil
			}
			return decimal.NewFromInt(math.MaxInt), nil
		}
```

File: pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go (hunk 17 above)
Comment (justinkaseman) [reply pkg/capabilities/consensus/ocr3/aggregators/reduce_aggregator.go]: Leaving for now, but I'll make a note in the ticket for Matt's suggestion of Field paths to get this in

**Review by cedric-cordenier — APPROVED** [review]

**Review by bolekk — APPROVED** [review]
